keps/sig-storage/20191125-bucket-provisioning.md

```diff
@@ -0,0 +1,428 @@
+1. `isDefaultBucketClass`: boolean. When true, signals that the controller should attempt to match `Bucket`’s without a defined `BucketClass` to this class, accounting for the `Bucket`’s requested protocol.  Multiple default classes for the same protocol will produce undefined behaviour.
+1. `supportedProtocols`: protocols the associated object store supports.  Applied when matching Bucket to BucketClasses. Admins may specify all protocols a provisioner supports.  `Bucket.spec.protocol` will be checked against the array prior to provisioning.
+1. `anonymousAccessMode`: (Optional) ACL specifying *uncredentialed* access to the Bucket.  This is applicable for cases where the storage instance or objects are intended to be publicly readable and/or writable.
+1. `additionalPermittedNamespaces`: A list of namespace *in addition to the originating namespace* that will be allowed access to this Bucket.  A nil and empty string value are equivalent and must not append any namespace to `Bucket.spec.permittedNamespaces` list.
+1. `releasePolicy`: Prescribes outcome of a Delete events. **Note:** In Brownfield and Static cases, *Retain* is mandated. 
+   - `Delete`:  the bucket and its contents are destroyed
+   - `Retain`:  the bucket and its data are preserved with only abstracting Kubernetes being destroyed
+1. `parameters`:  (Optional)  A map of string, string key values.  Allows admins to control user and access provisioning by setting provisioner key-values.
+
+### Access APIs
+
+The Access APIs abstract the backend policy system.  Access policy and user identities are an integral part of most object stores.  As such, a system must be implemented to manage both user/credential creation and the binding of those users to individual buckets via policies.  Object stores differ from file and block storage in how they manage users, with cloud providers typically integrating with an IAM platform.  This API includes support for cloud platform identity integration with Kubernetes ServiceAccounts.  On-prem solutions usually provide their own user management systems, which may look very different from each other and from IAM platforms.  We must also account for third party authentication solutions that may be integrated with an on-prem service.
+
+#### BucketAccessRequest
+
+A user facing API representing an object store user and an access policy defining the user’s relation to a storage instance.  A user will create a BucketAccessRequest in the app namespace.  A BucketAccessRequest can specify *either* a ServiceAccount in the same namespace or a desired Secret name.  Specifying a ServiceAccount enables provisioners to support cloud provider identity integration with their respective Kubernetes cluster offerings.
+
+```yaml
+apiVersion: cosi.io/v1alpha1
+kind: BucketAccessRequest
+metadata:
+  name:
+  namespace:
+  labels:
+    cosi.io/provisioner: [1]
+  finalizers:
+  - cosi.io/finalizer [2]
+spec:
+  serviceAccountName: [3]
+  bucketRequest: [4] 
```

Comment:
A BucketAccessRequest references a BucketRequest which may or may not need to exist? But this is a one-way link since a BucketRequest does _not_ reference a BucketAccessRequest.

Reply:
Lets discuss this further tomorrow

Reply:
Update: discussing with @wlan0 : the field name should be `bucket` not bucketRequest

Reply:
@wlan0 the field wil be named `bucket`, per above. Is this field omitted for greenfield and required for brownfield (like in provisioning)? I assume COSI will fill in this name in greenfield. In brownfield the admin, who has to create the `Bucket` will also manually create the BA and the user needs to know the name of the BA when she creates her BAR. Is this correct?

Reply:
It is possible to set it for the greenfield case as well, say when requesting to create a bucket by a certain name.

>  In brownfield the admin, who has to create the Bucket will also manually create the BA, and the user needs to know the name of the BA when she creates her BAR. Is this correct?

The credentials can be auto-provisioned even in the case of brownfield. This can be signaled by leaving the bucketAccessName empty.